include/ginkgo/core/base/math.hpp

```diff
@@ -314,35 +314,82 @@ struct next_precision_base_impl<std::complex<T>> {
 };
 
 
-#if GINKGO_ENABLE_HALF || GINKGO_ENABLE_BFLOAT16
+template <typename T, typename... Rest>
+struct next_precision_list_impl {};
 
+template <typename T, typename U, typename... Rest>
+struct next_precision_list_impl<T, U, Rest...> {
+    using type = typename next_precision_list_impl<T, Rest...>::type;
+};
+
+template <typename T, typename U, typename... Rest>
+struct next_precision_list_impl<T, T, U, Rest...> {
+    using type = U;
+};
 
 template <typename T>
-struct next_precision_impl {};
+struct next_precision_impl {
+    using type =
+        typename next_precision_list_impl<T,
+                                          double,  // put the last entry here
+                                                   // again to make the
+                                                   // searching easier
+#if GINKGO_ENABLE_HALF
+                                          half,
+#endif
+#if GINKGO_ENABLE_BFLOAT16
+                                          bfloat16,
+#endif
+                                          float, double>::type;
+};
 
 
-template <>
-struct next_precision_impl<gko::float16> {
-    using type = float;
+template <typename T>
+struct next_precision_impl<std::complex<T>> {
+    using type = std::complex<typename next_precision_impl<T>::type>;
 };
 
-template <>
-struct next_precision_impl<float> {
-    using type = double;
-};
 
-template <>
-struct next_precision_impl<double> {
-    using type = gko::float16;
+/**
+ * Move to next precision until move is zero
+ */
+template <typename T, unsigned move>
+struct next_precision_move_impl {
+    using type = typename next_precision_move_impl<
+        typename detail::next_precision_impl<T>::type, move - 1>::type;
 };
 
 template <typename T>
-struct next_precision_impl<std::complex<T>> {
-    using type = std::complex<typename next_precision_impl<T>::type>;
+struct next_precision_move_impl<T, 0> {
+    using type = T;
+};
+
+/**
+ * Move U until next_precision_move_impl<T, move> == U
+ */
+template <typename T, unsigned move,
+          typename U = typename next_precision_impl<T>::type>
```

Comment:
shouldn't it be 
```suggestion
          typename U = typename next_precision_impl<T, move>::type>
```

Reply:
no, it is just to start the search from the next one.
The stop criterion of recursion is when the searching U is the same type as T.
It might jump to much for example `<half, 2>` in `<half, float, double>`, the search will only check `double`, `half` (stop) without considering `float`.